use-cases/xtclim/src/trainer.py

```diff
@@ -14,56 +14,114 @@
 tracker.start()
 """
 
+from typing import Literal, Optional
 import torch
 import torch.optim as optim
 import numpy as np
 import pandas as pd
+import sys
+import os
+
+from itwinai.components import monitor_exec
+from itwinai.torch.trainer import TorchTrainer
+from itwinai.torch.distributed import (
+    distributed_resources_available,
+    TorchDistributedStrategy,
+    TorchDDPStrategy,
+    HorovodStrategy,
+    DeepSpeedStrategy,
+    NonDistributedStrategy
+)
+from itwinai.loggers import Logger
+from itwinai.torch.config import TrainingConfiguration
 
-from itwinai.components import Trainer, monitor_exec
 import model
 from torch.utils.data import DataLoader
 from torchvision.utils import make_grid
 from engine import train, validate
 from utils import save_reconstructed_images, save_loss_plot, save_ex
-from initialization import device, beta, criterion
+from initialization import beta, criterion, pixel_wise_criterion
 
-class TorchTrainer(Trainer):
+class XTClimTrainer(TorchTrainer):
     def __init__(
             self,
             epochs: int,
             batch_size: int,
-            lr: float
+            lr: float,
+            strategy: Literal["ddp", "deepspeed", "horovod"] = 'ddp',
+            save_best: bool = True,
+            logger: Optional[Logger] = None
     ):
-        super().__init__()
-        self.epochs = epochs
+        super().__init__(
+            epochs=epochs,
+            config={},
+            strategy=strategy,
+            logger=logger
+        )
         self.batch_size = batch_size
+        self.epochs = epochs
         self.lr = lr
+        # Global training configuration
+        self.config = TrainingConfiguration(
+            batch_size=self.batch_size,
+            save_best=save_best,
+        )
+
+    def final_loss(self, bce_loss, mu, logvar, beta=0.1):
+        """
+        Adds up reconstruction loss (BCELoss) and Kullback-Leibler divergence.
+        KL-Divergence = 0.5 * sum(1 + log(sigma^2) - mu^2 - sigma^2)
+
+        Parameters:
+        bce_loss: recontruction loss
+        mu: the mean from the latent vector
+        logvar: log variance from the latent vector
+        beta: weight over the KL-Divergence
+        """
+        BCE = bce_loss
+        KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
+        return BCE + beta*KLD
+
 
     @monitor_exec
     def execute(self):
```

Comment:
it would be nice to separate the boilerplate code (e.g., distributed strategy setup, loggers setup) from the actual training code, if possible. Ideally, the use cases should not worry about the setup and should only override the `train` method from the `TorchTrainer` class. In other words, use cases can get the best out of the `TorchTrainer` when reusing its `execute` method and overriding only the `train` method

Reply:
As we had discussed, in this case, the preprocessing part first writes the files to disk and then the trainer reads the data from disk. Hence `execute` is not used, rather `@monitor_exec` is employed.

Reply:
I have similar comments as for the `execute` method in the `XTClimPredictor`